spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/Tokenizer.java

```diff
@@ -49,80 +46,83 @@
-        .getSchema().getType();
-      Preconditions.checkArgument(columnToBeTokenizedType == Schema.Type.STRING,
-                                  "Column to be tokenized %s must be of type String, but was of type %s.",
-                                  config.columnToBeTokenized, columnToBeTokenizedType);
+      Schema schema = inputSchema.getField(config.columnToBeTokenized).getSchema();
+      Schema.Type type = schema.isNullable() ? schema.getNonNullable().getType() : schema.getType();
+      Preconditions.checkArgument(type == Schema.Type.STRING, "Column to be tokenized %s must be of type String, " +
+        "but was of type %s.", config.columnToBeTokenized, type);
     }
   }
 
   @Override
+  public void initialize(SparkExecutionPluginContext context) throws Exception {
+    super.initialize(context);
+    tokenizer = new RegexTokenizer().setInputCol(config.columnToBeTokenized).setOutputCol(config.outputColumn)
+      .setPattern(config.patternSeparator);
+  }
+
+  @Override
   public JavaRDD<StructuredRecord> transform(SparkExecutionPluginContext context,
                                              JavaRDD<StructuredRecord> input) throws Exception {
     JavaSparkContext javaSparkContext = context.getSparkContext();
     SQLContext sqlContext = new SQLContext(javaSparkContext);
-    //Create outputschema
-    outputSchema = Schema.recordOf("outputSchema", Schema.Field.of(config.outputColumn,
-                                                                   Schema.arrayOf(Schema.of(Schema.Type.STRING))));
-    //Schema to be used to create dataframe
-    StructType schema = new StructType(new StructField[]{
-      new StructField(config.columnToBeTokenized, DataTypes.StringType, false, Metadata.empty())});
-    //Transform input i.e JavaRDD<StructuredRecord> to JavaRDD<Row>
-    JavaRDD<Row> rowRDD = input.map(new Function<StructuredRecord, Row>() {
+    if (input == null) {
+      return context.getSparkContext().emptyRDD();
+    }
+    outputSchema = outputSchema != null ? outputSchema : config.getOutputSchema(input.first().getSchema(),
+                                                                                config.outputColumn);
+    JavaRDD<String> javardd = input.map(new Function<StructuredRecord, String>() {
       @Override
-      public Row call(StructuredRecord rec) throws Exception {
-        return RowFactory.create(rec.get(config.columnToBeTokenized));
+      public String call(StructuredRecord structuredRecord) throws Exception {
+        return StructuredRecordStringConverter.toJsonString(structuredRecord);
       }
     });
-    DataFrame sentenceDataFrame = sqlContext.createDataFrame(rowRDD, schema);
-    RegexTokenizer tokenizer = new RegexTokenizer().setInputCol(config.columnToBeTokenized)
-      .setOutputCol(config.outputColumn).setPattern(config.pattern);
+    DataFrame sentenceDataFrame = sqlContext.read().json(javardd);
```

Comment:
all of this is really overkill to do a simple tokenization. We are converting each record to a json string, converting that RDD into a DataFrame and parsing the json back into columns, tokenizing on the data frame, then transforming it back to json then back to an RDD of records.

Instead, just do something like:

```
  @Override
  public void initialize(SparkExecutionPluginContext context) throws Exception {
    super.initialize(context);
    separator = Pattern.compile(conf.separatorPattern);
  }

  public JavaRDD<StructuredRecord> transform(SparkExecutionPluginContext context,
                                             JavaRDD<StructuredRecord> input) throws Exception {
      return input.map(new Function<StructuredRecord, StructuredRecord>() {
        @Override
        public StructuredRecord call(StructuredRecord inputRecord) throws Exception {
          String val = inputRecord.get(conf.columnToBeTokenized);

          StructuredRecord.Builder outputBuilder = StructuredRecord.builder(outputSchema);
          for (Schema.Field field : inputRecord.getSchema().getFields()) {
            outputBuilder.set(field.getName(), inputRecord.get(field.getName()));
          }

          List<String> tokens = new ArrayList<>();
          if (val != null) {
            for (String token : separator.split(val)) {
              tokens.add(token);
            }
          }
          outputBuilder.set(conf.outputColumn, tokens);

          return outputBuilder.build();
        }
      });
  }
```

In fact, you can just do whatever is inside the map function in a Transform instead of a SparkCompute.